lib/Onyx.js

```diff
@@ -428,17 +475,15 @@ function set(key, val) {
  * @returns {Promise}
  */
 function multiSet(data) {
-    // AsyncStorage expenses the data in an array like:
-    // [["@MyApp_user", "value_1"], ["@MyApp_key", "value_2"]]
-    // This method will transform the params from a better JSON format like:
-    // {'@MyApp_user': 'myUserValue', '@MyApp_key': 'myKeyValue'}
-    const keyValuePairs = _.reduce(data, (finalArray, val, key) => ([
-        ...finalArray,
-        [key, JSON.stringify(val)],
-    ]), []);
+    const keyValuePairs = prepareKeyValuePairsForStorage(data);
+
+    // Capture (non-stringified) changes to cache
+    cache.merge(_.pairs(data));
+
+    // Optimistically inform subscribers
+    _.each(data, (val, key) => keyChanged(key, val));
 
     return AsyncStorage.multiSet(keyValuePairs)
-        .then(() => _.each(data, (val, key) => keyChanged(key, val)))
         .catch(error => evictStorageAndRetry(error, multiSet, data));
```

Comment:
> Not sure how we want to handle this. But if `evictStorageAndRetry()` is called then once it's done removing stuff it will call `multiSet()` again and also update the cache + call `keyChanged()` again.

Calling subscribers with the same data again isn't a big problem if any, at most they will re-render

But I see a bigger issue: This means that there is an existing concurrency issue even before my changes
1. multiSet is called to write some data
2. multiSet fails and schedules a retry - evictStorageAndRetry
3. another call to multiSet arrives before the retry is performed. It sets a new value for some of the keys
4. retry runs and overwrites the new value with the old value

The above is true for every method that currently uses `evictStorageAndRetry` to set up a retry

I've talked about a write queue - It would be best to implement one, but that's a separate issue

Methods that persist data to file should not write directly but push a task to a queue and ask the "queue manager" to flush the queue - if a flush is not running it will start. Flush will use the queue to write data to file - FIFO and handle any necessary retires. In the mean cache would always provide the most up to date data since it doesn't have to wait for the thing to be written to disk 

Furthermore having such a queue would allow you to group and set multiple keys at once (there is still some bridge flooding due to many writes running at the same time)
It would also allow you to skip outdated writes. Let's say something is updated very often in short intervals of time e.g. key "A" value "V" then again key "A" value "Val" and finally key "A" value "Value" - you don't need to first persist "V" then "Val" you can directly persist the latest available value - "Value"

Reply:
>Calling subscribers with the same data again isn't a big problem if any, at most they will re-render

Cool. I'm not sure if it's a problem just noting the behavior change. It seems like if we could avoid it then it would be nice to do so. One alternative I'd propose is that we make the "retryable" part of [`evictStorageAndRetry`](https://github.com/Expensify/react-native-onyx/blob/0cdf133867dccbd0b99090850130fef3442e826f/lib/Onyx.js#L394-L407
) only operate on the data in storage and not manipulate the cache at all. That would solve the problem in another way.

>But I see a bigger issue: This means that there is an existing concurrency issue even before my changes

Sounds like an interesting future improvement/observation, but I'm not sure this is a problem we are experiencing atm. Usually what causes the storage to throw errors is that a very large `reportActions_*` set is processed and merged. Those are usually merged in and not written with `multiSet()` or `set()`.